synapse/federation/sender/per_destination_queue.py

```diff
@@ -415,22 +416,93 @@ async def _catch_up_transmission_loop(self) -> None:
                     "This should not happen." % event_ids
                 )
 
-            if logger.isEnabledFor(logging.INFO):
-                rooms = [p.room_id for p in catchup_pdus]
-                logger.info("Catching up rooms to %s: %r", self._destination, rooms)
+            # We send transactions with events from one room only, as its likely
+            # that the remote will have to do additional processing, which may
+            # take some time. It's better to give it small amounts of work
+            # rather than risk the request timing out and repeatedly being
+            # retried, and not making any progress.
+            for pdu in catchup_pdus:
+                # The PDU from the DB will be the last PDU in the room from
+                # *this server* that wasn't sent to the remote. However, other
```

Comment:
This is because we only know that *we* failed to send an event, but the remote server might be communicating OK with other servers in the room? (Meaning that it could actually already have the event we're trying to send to it?)

Reply:
Err, I don't follow really. The server will only track the last event it failed to send to the remote was, which will only be local events. When the server comes back up we want it to be told about the most recent events in the room, so that it can figure out it missed stuff. We don't want to send anything other than the latest events as that will cause unnecessary processing on the remote, and the last event sent by the server to a room won't necessarily be the part of the latest events in the room (as other servers may have sent events since).

In other words, we want the remote to only receive the current forward extremities in the room; we're not trying to send it all events its missed.

Reply:
> The server will only track the last event it failed to send to the remote was, which will only be local events.

This is what I was trying to confirm. 👍